src/us/kbase/auth2/lib/storage/mongo/MongoStorage.java

```diff
@@ -721,19 +744,54 @@ private Document getUserDoc(
 			final boolean local)
 			throws AuthStorageException, NoSuchUserException {
 		requireNonNull(userName, "userName");
-		final Document projection = new Document(Fields.USER_PWD_HSH, 0)
-				.append(Fields.USER_SALT, 0);
-		final Document user = findOne(collection,
-				new Document(Fields.USER_NAME, userName.getName()), projection);
+		Document user = getUserDocWithoutPassword(collection, userName);
 		if (user == null) {
 			throw new NoSuchUserException(userName.getName());
 		}
+		if (user.getString(Fields.USER_ANONYMOUS_ID) == null) {
+			user = updateUserAnonID(collection, userName);
+		}
 		if (local && !user.getBoolean(Fields.USER_LOCAL)) {
 			throw new NoSuchLocalUserException(userName.getName());
 		}
 		return user;
 	}
 
+	private Document getUserDocWithoutPassword(final String collection, final UserName userName)
+			throws AuthStorageException {
+		return findOne(
+				collection,
+				new Document(Fields.USER_NAME, userName.getName()),
+				new Document(Fields.USER_PWD_HSH, 0).append(Fields.USER_SALT, 0));
+	}
+
+	private Document updateUserAnonID(final String collection, final UserName userName)
+			throws AuthStorageException {
+		/* Added in version 0.6.0 when anonymous IDs were added to users. This method lazily
+		 * backfills the anonymous ID on an as-needed basis.
+		 */
+		final Document query = new Document(Fields.USER_NAME, userName.getName())
+				// only modify if not already done to avoid race conditions
+				.append(Fields.USER_ANONYMOUS_ID, null);
+		final Document update = new Document(
+				"$set", new Document(Fields.USER_ANONYMOUS_ID, randGen.randomUUID().toString()));
+		try {
+			db.getCollection(COL_USERS).updateOne(query, update);
+			// if it didn't match we assume the user doc was updated to include an ID in
+			// a different thread or process
```

Comment:
> // if it didn't match we assume the user doc was updated to include an ID in
> // a different thread or process

Is this a possibility here? Is there a way to test or double check?
Although, I guess as long as the document that gets returned at the end of the function has an anon ID, it's fine.

Reply:
> Is this a possibility here?

It's possible - 2 threads pull the document, see there's no anon ID, generate different UUIDs, and one will beat the other in the database update.

> Is there a way to test or double check?

The test that calls this method via reflection with a user in the database with a anon ID in place tests this scenario.
https://github.com/kbase/auth2/pull/400/files#diff-309afd3d9854b938220b5230c21e238e104c3c176be56807f8757feb33518f83R117-R120

> Although, I guess as long as the document that gets returned at the end of the function has an anon ID, it's fine.

... unless the anon ID isn't the same as the anon ID in the database. The various protections in this method are designed to prevent that from happening.